// eslint-config-es/build/lib/rules/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unicorn = exports.typescript = exports.react = exports.mochaRules = exports.extended = exports.createSharedRulesFor = exports.coreRules = exports.comments = void 0;
const comments_1 = require("./comments");
Object.defineProperty(exports, "comments", { enumerable: true, get: function () { return comments_1.comments; } });
const core_1 = require("./core");
Object.defineProperty(exports, "coreRules", { enumerable: true, get: function () { return core_1.coreRules; } });
const sharedCoreTypescript_1 = require("./sharedCoreTypescript");
Object.defineProperty(exports, "createSharedRulesFor", { enumerable: true, get: function () { return sharedCoreTypescript_1.createSharedRulesFor; } });
const extended_1 = require("./extended");
Object.defineProperty(exports, "extended", { enumerable: true, get: function () { return extended_1.extended; } });
const mocha_1 = require("./mocha");
Object.defineProperty(exports, "mochaRules", { enumerable: true, get: function () { return mocha_1.mochaRules; } });
const react_1 = require("./react");
Object.defineProperty(exports, "react", { enumerable: true, get: function () { return react_1.react; } });
const typescript_1 = require("./typescript");
Object.defineProperty(exports, "typescript", { enumerable: true, get: function () { return typescript_1.typescript; } });
const unicorn_1 = require("./unicorn");
Object.defineProperty(exports, "unicorn", { enumerable: true, get: function () { return unicorn_1.unicorn; } });
