// solidbench/lib/Templates.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Templates = void 0;
const path_1 = __importDefault(require("path"));
/**
 * Paths to default configuration files.
 */
exports.Templates = {
    ENHANCEMENT_CONFIG: path_1.default.join(__dirname, '../templates/enhancer-config-pod.json'),
    FRAGMENT_CONFIG: path_1.default.join(__dirname, '../templates/fragmenter-config-pod.json'),
    ENHANCEMENT_FRAGMENT_CONFIG: path_1.default.join(__dirname, '../templates/fragmenter-auxiliary-config-pod.json'),
    QUERY_CONFIG: path_1.default.join(__dirname, '../templates/query-config.json'),
    SERVER_CONFIG: path_1.default.join(__dirname, '../templates/server-config.json'),
    QUERIES_DIRECTORY: path_1.default.join(__dirname, '../templates/queries'),
    VALIDATION_CONFIG: path_1.default.join(__dirname, '../templates/validation-config.json'),
    VALIDATION_PARAMS_URL: 'https://cloud.ilabt.imec.be/index.php/s/bBZZKb7cP95WgcD/download/validation_params.zip',
};
