// ldbc-snb-enhancer/lib/parameters/ParameterEmitterCsv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParameterEmitterCsv = void 0;
const fs = require("fs");
/**
 * Emits parameters as CSV files.
 */
class ParameterEmitterCsv {
    constructor(destinationPath) {
        this.headerLength = -1;
        this.destinationPath = destinationPath;
        this.fileStream = fs.createWriteStream(this.destinationPath);
    }
    emitHeader(columnNames) {
        // Validate columns
        if (this.headerLength === -1) {
            this.headerLength = columnNames.length;
        }
        else {
            throw new Error('Attempted to emit header more than once.');
        }
        this.emitRow(columnNames);
    }
    emitRow(columns) {
        // Validate columns
        if (columns.length !== this.headerLength) {
            throw new Error(`A column of length ${columns.length} was emitted, while length ${this.headerLength} is required.`);
        }
        this.fileStream.write(`${columns.join(',')}\n`);
    }
    flush() {
        this.fileStream.end();
    }
}
exports.ParameterEmitterCsv = ParameterEmitterCsv;
